### PR TITLE
LatteExtension: added support for macro factories defined as DI services

### PR DESCRIPTION
**Pull request by hrach**
- feature
- issues - no
- documentation - https://github.com/nette/web-content/pull/470
- BC break - no

### CODE REVIEW

**Comment by JanTvrdik** [comment]
> but I can't see 2.5/master

@dg will probably put this to 2.4 anyway

**Comment by Majkl578** [comment]
It probably should support `@macroFactory::customInstall" format as well to reflect current behavior.

**Comment by dg** [comment]
You can send documentation PR to 2.4, it is defacto master.

**Comment by dg** [comment]
@hrach @Majkl578  I have not studied this PR deeply, but I see in description is *BC break - no* and v2.5 will not be released until next year, so can you explain `:-/` to me?

**Comment by hrach** [comment]
That was just my idea about semantic versioning, which you/nette doesn't use, so probably the point about 2.5 is useless and it's my turn - to prepare the doc & maybe implement Majkl's request.

**Comment by dg** [comment]
Ok. (Of course I can label it as `2.5 milestone` and merge it in few months when branch 2.5 will be opened, just tell me.)

**Comment by hrach** [comment]
I tried to find place where should I add the doc, but I failed. Could you please point me out to the proper page?

**Comment by hrach** [comment]
Reimplemented with Majk's suggestion. Still, I have no idea where to put the doc for such behavior. cc @dg

**Comment by dg** [comment]
Do you mean which branch or file?

**Comment by hrach** [comment]
File.

**Comment by dg** [comment]
Probably this file https://github.com/nette/web-content/blob/doc-2.4/cs/configuring.texy

**Comment by hrach** [comment]
Great, will add it there today :-)

File: src/Bridges/ApplicationDI/LatteExtension.php [hunk 1]
```diff
@@ -71,14 +68,28 @@ public function loadConfiguration()
...
+	public function addMacro($macro)
 	{
 		$builder = $this->getContainerBuilder();
-		$builder->getDefinition($this->prefix('latteFactory'))
-			->addSetup('?->onCompile[] = function ($engine) { ' . $macro . '($engine->getCompiler()); }', ['@self']);
+		$definition = $builder->getDefinition($this->prefix('latteFactory'));
+
+		if (isset($macro[0]) && $macro[0] === '@') {
+			if (strpos($macro, '::') === FALSE) {
+				$method = 'install';
```
Comment (dg) [inline src/Bridges/ApplicationDI/LatteExtension.php]: I think that auto-adding `install` should be omitted in this case.

**Comment by TomasVotruba** [comment]
What needs to be done here?

**Comment by TomasVotruba** [comment]
@hrach Could you make that change please?

File: src/Bridges/ApplicationDI/LatteExtension.php (hunk 1 above)
Comment (hrach) [inline src/Bridges/ApplicationDI/LatteExtension.php]: I'm ok with compulsory method definition in config.

File: src/Bridges/ApplicationDI/LatteExtension.php [hunk 2]
```diff
@@ -71,14 +68,28 @@ public function loadConfiguration()
...
+			} else {
+				list($macro, $method) = explode('::', $macro);
+			}
+			$definition->addSetup('?->onCompile[] = function ($engine) { ?->' . $method . '($engine->getCompiler()); }', ['@self', $macro]);
+
+		} else {
+			if (strpos($macro, '::') === FALSE && class_exists($macro)) {
```
Comment (dg) [inline src/Bridges/ApplicationDI/LatteExtension.php]: Why is here `class_exists`?

Comment (hrach) [inline src/Bridges/ApplicationDI/LatteExtension.php]: This is only moved code which was here before the PR.

Comment (dg) [inline src/Bridges/ApplicationDI/LatteExtension.php]: :-)

**Comment by dg** [comment]
Thanks!

**Comment by hrach** [comment]
:+1: Great, more DI to our code (actually, before this it was not possible to write more complex macros with DI (aka injected LinkService)).

**Pull request merged**